// @herb-tools/linter/dist/src/rules/html-no-underscores-in-attribute-names.js
import { ParserRule } from "../types.js";
import { AttributeVisitorMixin } from "./rule-utils.js";
import { getStaticContentFromNodes } from "@herb-tools/core";
import { IdentityPrinter } from "@herb-tools/printer";
class HTMLNoUnderscoresInAttributeNamesVisitor extends AttributeVisitorMixin {
    checkStaticAttributeStaticValue({ attributeName, attributeNode }) {
        this.check(attributeName, attributeNode);
    }
    checkStaticAttributeDynamicValue({ attributeName, attributeNode }) {
        this.check(attributeName, attributeNode);
    }
    checkDynamicAttributeStaticValue({ nameNodes, attributeNode }) {
        const attributeName = getStaticContentFromNodes(nameNodes);
        this.check(attributeName, attributeNode);
    }
    checkDynamicAttributeDynamicValue({ nameNodes, attributeNode }) {
        const attributeName = getStaticContentFromNodes(nameNodes);
        this.check(attributeName, attributeNode);
    }
    check(attributeName, attributeNode) {
        if (!attributeName)
            return;
        if (attributeName.includes("_")) {
            this.addOffense(`Attribute \`${IdentityPrinter.print(attributeNode.name)}\` should not contain underscores. Use hyphens (-) instead.`, attributeNode.value.location, "warning");
        }
    }
}
export class HTMLNoUnderscoresInAttributeNamesRule extends ParserRule {
    name = "html-no-underscores-in-attribute-names";
    check(result, context) {
        const visitor = new HTMLNoUnderscoresInAttributeNamesVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
