// @herb-tools/linter/dist/src/rules/html-boolean-attributes-no-value.js
import { ParserRule } from "../types.js";
import { AttributeVisitorMixin, isBooleanAttribute, hasAttributeValue } from "./rule-utils.js";
import { IdentityPrinter } from "@herb-tools/printer";
class BooleanAttributesNoValueVisitor extends AttributeVisitorMixin {
    checkStaticAttributeStaticValue({ originalAttributeName, attributeNode }) {
        this.checkAttribute(originalAttributeName, attributeNode);
    }
    checkStaticAttributeDynamicValue({ originalAttributeName, attributeNode }) {
        this.checkAttribute(originalAttributeName, attributeNode);
    }
    checkAttribute(attributeName, attributeNode) {
        if (!isBooleanAttribute(attributeName))
            return;
        if (!hasAttributeValue(attributeNode))
            return;
        this.addOffense(`Boolean attribute \`${IdentityPrinter.print(attributeNode.name)}\` should not have a value. Use \`${attributeName.toLowerCase()}\` instead of \`${IdentityPrinter.print(attributeNode)}\`.`, attributeNode.value.location, "error");
    }
}
export class HTMLBooleanAttributesNoValueRule extends ParserRule {
    name = "html-boolean-attributes-no-value";
    check(result, context) {
        const visitor = new BooleanAttributesNoValueVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
